UCP/RNDV:Add put_zcopy in UCP RNDV protocol.

xinzhao3 opened the pull request.

bureddy commented:
These changes are in a direction to use PUT protocol in cases where GET(READ) protocol is not efficient in GPU Direct RDMA protocols

mellanox-github commented:
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4812/ for details (Mellanox internal link).

swx-jenkins commented:
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2806/ for details.

evgeny-leksikov commented on src/ucp/core/ucp_context.c:
```diff
@@ -93,6 +93,11 @@ static ucs_config_field_t ucp_config_table[] = {
+  {"USE_RNDV_PUT_ZCOPY", "n", "Use put_zcopy in RNDV protocol.\n"
+   "n - Do not use put_zcopy in RNDV protocol, use default get_zcopy"
+   "y - Use put_zcopy in RNDV protocol",
+   ucs_offsetof(ucp_config_t, ctx.use_rndv_put_zcopy), UCS_CONFIG_TYPE_BOOL},
```
I think context is too global place for protocol switching. Maybe better to make the decision per EP?

evgeny-leksikov commented on src/ucp/tag/rndv.c:
```diff
@@ -585,6 +695,97 @@ UCS_PROFILE_FUNC(ucs_status_t, ucp_rndv_progress_bcopy_send, (self),
+                         ucp_ep_config(rndv_req->send.ep)->tag.rndv.max_put_zcopy);
+    }
+
+    iov[0].buffer = (void*)rndv_req->send.buffer + offset;
```
Pls use `ucp_dt_iov_copy_uct`

evgeny-leksikov commented on src/ucp/tag/rndv.c:
```diff
@@ -585,6 +695,97 @@ UCS_PROFILE_FUNC(ucs_status_t, ucp_rndv_progress_bcopy_send, (self),
+    iov[0].memh   = rndv_req->send.state.dt.contig.memh;
+    iov[0].count  = 1;
+    iov[0].stride = 0;
+    rndv_req->send.uct_comp.count++;
```
Pls use `ucp_request_send_state_advance`. it will simplify logic below

evgeny-leksikov commented on src/ucp/tag/rndv.c:
```diff
@@ -689,6 +890,29 @@ UCS_PROFILE_FUNC(ucs_status_t, ucp_rndv_rtr_handler,
+        rndv_req->send.buffer         = sreq->send.buffer;
+        rndv_req->send.length         = sreq->send.length;
+        rndv_req->send.uct_comp.func  = ucp_rndv_put_completion;
+        rndv_req->send.uct_comp.count = 0;
```
Pls use `ucp_request_send_state_reset`

evgeny-leksikov commented on src/ucp/tag/rndv.c:
```diff
@@ -67,6 +68,30 @@ size_t ucp_tag_rndv_pack_rkey(ucp_request_t *sreq, ucp_lane_index_t lane,
+     * is its datatype contiguous and does the receive side need it */
+    if (ucp_ep_rndv_md_flags(ep) & UCT_MD_FLAG_NEED_RKEY) {
+        status = ucp_request_recv_buffer_reg(rreq, ep, lane);
+        ucs_assert_always(status == UCS_OK);
```
Can we propagate the error to user level?

evgeny-leksikov commented on src/ucp/tag/rndv.c:
```diff
@@ -585,6 +695,97 @@ UCS_PROFILE_FUNC(ucs_status_t, ucp_rndv_progress_bcopy_send, (self),
+
+        /* switch to AM */
+        if ((UCP_DT_IS_CONTIG(rndv_req->send.datatype)) &&
+            (rndv_req->send.length >= ucp_ep_config(rndv_req->send.ep)->am.zcopy_thresh[0])) {
```
Can `ucp_proto_get_zcopy_threshold` be useful here?

alinask commented on src/ucp/tag/rndv.c:
```diff
@@ -426,6 +517,11 @@ UCS_PROFILE_FUNC_VOID(ucp_rndv_matched, (worker, rreq, rndv_rts_hdr),
             ucp_rndv_handle_recv_contig(rndv_req, rreq, rndv_rts_hdr);
         } else {
+            if (ep->worker->context->config.ext.use_rndv_put_zcopy &&
+                ucp_ep_is_rndv_lane_present(ep)) {
```
maybe worth checking here that (rndv_rts_hdr->address != 0)

alinask commented:
was this tested at scale?
is there a gtest that covers put_zcopy with rndv?

xinzhao3 commented:
@alinask it is not tested with scale yet, will test asap. There is no gtest cover this put_zcopy. Will add one.

alinask commented:
@xinzhao3 thanks! most of the rndv tests in gtest are in the test_ucp_tag_xfer.cc file.
It has many combinations of datatypes and thresholds to cover many cases.

xinzhao3 commented:
bot:retest

xinzhao3 commented:
@evgeny-leksikov @alinask could you review the pr again? I added test cases to gtest, also test this branch up to 512 processes.

swx-jenkins commented:
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2948/ for details.

swx-jenkins commented:
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2949/ for details.

swx-jenkins commented:
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2950/ for details.

mellanox-github commented:
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4994/ for details (Mellanox internal link).

mellanox-github commented:
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4995/ for details (Mellanox internal link).

mellanox-github commented:
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4996/ for details (Mellanox internal link).

evgeny-leksikov commented on src/ucp/tag/rndv.c:
```diff
@@ -574,6 +686,90 @@ UCS_PROFILE_FUNC(ucs_status_t, ucp_rndv_progress_bcopy_send, (self),
+            uct_rkey_release(&rndv_req->send.rndv_put.rkey_bundle);
+        }
+
+        /* switch to AM */
```
Is it possible to choose the exact protocol by thresholds calculation before stating the request?

swx-jenkins commented:
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2970/ for details.

mellanox-github commented:
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5031/ for details (Mellanox internal link).

swx-jenkins commented:
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2977/ for details.

swx-jenkins commented:
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2978/ for details.

swx-jenkins commented:
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2979/ for details.

xinzhao3 commented:
@evgeny-leksikov @alinask could you review again? I changed to use environment variable: get_zcopy, put_zcopy, auto. Currently default is get_zcopy, because this PR is orthogonal to CUDA changes. When CUDA changes is added, I will make auto as default option.

mellanox-github commented:
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5038/ for details (Mellanox internal link).

mellanox-github commented:
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5039/ for details (Mellanox internal link).

xinzhao3 commented:
bot:retest

swx-jenkins commented:
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2986/ for details.

mellanox-github commented:
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5049/ for details (Mellanox internal link).

yosefe commented:
@alinask pls take a look

evgeny-leksikov commented on src/ucp/core/ucp_context.c:
```diff
@@ -36,6 +36,13 @@ static const char * ucp_device_type_names[] = {
+static const char * ucp_rndv_modes[] = {
+    [UCP_RNDV_MODE_GET_ZCOPY] = "get_zcopy",
+    [UCP_RNDV_MODE_PUT_ZCOPY] = "put_zcopy",
+    [UCP_RNDV_MODE_AUTO]      = "auto",
```
I do not see where "auto" value is handled.. Below, there are checks in if-statements to exact "put" or "get" values with 'else' cases. So, if user defines this value, behaviour may be erroneous, right? Pls, add a test for this value too.

swx-jenkins commented:
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/3016/ for details.

swx-jenkins commented:
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/3017/ for details.

swx-jenkins commented:
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/3018/ for details.

mellanox-github commented:
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5089/ for details (Mellanox internal link).

mellanox-github commented:
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5090/ for details (Mellanox internal link).

mellanox-github commented:
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5091/ for details (Mellanox internal link).

mellanox-github commented:
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5092/ for details (Mellanox internal link).

swx-jenkins commented:
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/3019/ for details.

swx-jenkins commented:
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/3020/ for details.

alinask commented on src/ucp/tag/rndv.c:
```diff
@@ -66,6 +66,30 @@ size_t ucp_tag_rndv_pack_rkey(ucp_request_t *sreq, ucp_lane_index_t lane,
+        status = ucp_request_recv_buffer_reg(rreq, ep, lane);
+        ucs_assert_always(status == UCS_OK);
+
+        /* if the send buffer was registered, send the rkey */
```
_send_ buffer?

alinask commented on src/ucp/tag/rndv.c:
```diff
@@ -574,6 +693,71 @@ UCS_PROFILE_FUNC(ucs_status_t, ucp_rndv_progress_bcopy_send, (self),
+    ucs_trace_data("ep: %p try to progress put_zcopy for rndv get. rndv_req: %p. lane: %d",
+                   rndv_req->send.ep, rndv_req, rndv_req->send.lane);
+
+    /* rndv_req is the internal request to perform the get operation */
```
_put_ operation

alinask commented on src/ucp/core/ucp_request.h:
```diff
@@ -124,6 +126,13 @@ struct ucp_request {
                 struct {
+                    uint64_t      remote_address; /* address of the sender's data buffer */
+                    uintptr_t     remote_request; /* pointer to the sender's send request */
+                    uct_rkey_bundle_t rkey_bundle;
```
shouldn't these be _receiver's_ buffer and request?

swx-jenkins commented:
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/3098/ for details.

mellanox-github commented:
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5192/ for details (Mellanox internal link).

swx-jenkins commented:
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/3099/ for details.

swx-jenkins commented:
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/3100/ for details.

swx-jenkins commented:
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/3101/ for details.

swx-jenkins commented:
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/3102/ for details.

mellanox-github commented:
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5193/ for details (Mellanox internal link).

mellanox-github commented:
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5194/ for details (Mellanox internal link).

mellanox-github commented:
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5195/ for details (Mellanox internal link).

mellanox-github commented:
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5196/ for details (Mellanox internal link).

swx-jenkins commented:
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/3103/ for details.

mellanox-github commented:
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5197/ for details (Mellanox internal link).

xinzhao3 commented:
bot:retest

swx-jenkins commented:
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/3104/ for details.

xinzhao3 commented:
@alinask @evgeny-leksikov I addressed all the comments, could you review the PR again?
There is one failure (malloc_hook_cplusplus.mmmap_ptrs) in Jenkins MellanoxLab, I think it is irrelevant with this PR.

mellanox-github commented:
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5198/ for details (Mellanox internal link).

evgeny-leksikov commented on src/ucp/tag/rndv.c:
```diff
@@ -79,15 +103,16 @@ static size_t ucp_tag_rndv_rts_pack(void *dest, void *arg)
-    if (UCP_DT_IS_CONTIG(sreq->send.datatype)) {
+    if (UCP_DT_IS_CONTIG(sreq->send.datatype) &&
+        (ep->worker->context->config.ext.rndv_mode == UCP_RNDV_MODE_GET_ZCOPY ||
+         ep->worker->context->config.ext.rndv_mode == UCP_RNDV_MODE_AUTO)) {
```
Does current implementation assume `AUTO` == `GET_ZCOPY`? maybe better to handle this on initialization? I mean 
```
if (rndv_mode == UCP_RNDV_MODE_AUTO) {
    rndv_mode = UCP_RNDV_MODE_GET_ZCOPY;
}
```

swx-jenkins commented:
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/3112/ for details.

mellanox-github commented:
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5205/ for details (Mellanox internal link).

xinzhao3 commented:
@evgeny-leksikov yeah currently AUTO == GET_ZCOPY. I changed to set it in initialization, could you review again?
